backend/routes/userNews.js

```diff
@@ -161,12 +173,104 @@ router.put("/:newsId/bookmarked", async (req, res) => {
     }
+
     res.json(personalNews);
   } catch (error) {
     res.status(500).json({ error: "Error updating metadata" });
   }
 });
 
+// get a personalized feed
+router.post("/personalized", async (req, res) => {
+    const newNews = await prisma.userNewsCache.findMany({
+      where: { userId: req.session.userId },
+      orderBy: {
+        id: "desc",
+      },
+    });
+
+    const personalNews = [];
+    // there are entries found
+    for (const art of newNews) {
+      const newArticle = await prisma.news.findMany({
+        where: { id: art.newsId },
+      });
+
+      newArticle[0].score = art.score;
+      newArticle[0].bookmarked = art.bookmarked;
+
+      personalNews.push(newArticle[0]);
+    }
+
+    try {
+         // sort news based on engagement scores
+    personalNews.sort((a, b) =>  (b['score']) - (a['score']))
+    const top10 = personalNews.slice(0,10); // top 10 articles with highest engagement scores
+
+    let categories = []   // aggregate the categories present in the top 10
+    for (const article of top10) {
+        for (const category of article.category) {
+            if (!categories.includes(category)) {
+                categories.push(category)
+            }
+        }
+    }
+
+    // perform similiarity analysis
+    weightedScores = {} // scores for every category (how many points a news article gets for having this category)
+    for (article of top10) {
+        const similiarity = getSimiliarity(categories, article.category)
+        for (let category of article.category) {
+            if (category in weightedScores) { // the category already exists
+                weightedScores[category] += similiarity; // add onto the exisiting value
+            } else {
+                weightedScores[category] = similiarity;
+            }
+        }
+    }
+    
+    // use the weights to give all news a score
+    const allNews = await prisma.news.findMany();
+    const usedNews = newNews.map(article => article.newsId)
+    const notUsedNews = allNews.filter(id => !usedNews.includes(id)); // so that we are not looking at articles that have already been seen
+
+    let rankings = []; // json of the newsIds and the rankings
+    for (const article of notUsedNews) {
+        // calculate the score
+        let score = 0.0;
+        for (const category of article.category) {
+            if (weightedScores.hasOwnProperty(category)) { // article has one of the weighted categories
+                score += weightedScores[category];
+            }
+        }
+        const newRanking = {
+            newsId : article.id,
+            totalScore : score,
+        }
+
+        rankings.push(newRanking)
+   }
+
+    // sort rankings for the articles that are the best choices
+    let sortedRankings = rankings.slice().sort((a, b) =>  (b['totalScore']) - (a['totalScore'])) // first 30 rankings (newsIds)
+    sortedRankings = sortedRankings.slice(0, 30).map(article => article.newsId);
```

Comment:
For the next iteration, we also want to improve this algorithm here to more efficiently find the top 30 recommendations